README.md

```diff
@@ -49,7 +51,48 @@ scripts/evaluate_agents.py agent1_file.py [agent2_file.py negamax random]
 
 `q_learning/` has an example written by the repo author that is runnable with the `evaluate_agents.py` script, but I have not yet verified that it will be runnable in a proper kaggle submission (the main uncertainty is whether the model state dict .pt file is submittable, as I have seen some examples where the state dict is written out as a variable within the function. I hope this won't be necessary, as it seems like a big pain!
 
-# Leaderboard
+# Leaderboard [Depracated?]
 Negamax (Kaggle) vs `q_learning` (j-bernardi): 0.0 : 1.0
 
 ^ Should be easy to beat me!
+
+# TrueSkill leaderboard
+A leaderboard is a good way to represent the performance of 2 or more agents.
+Kaggle has built their leaderboard using a variant of [TrueSkill](https://www.microsoft.com/en-us/research/project/trueskill-ranking-system/), a kind of Elo algorithm developed by Microsoft:
+> Our ranking system takes more inspiration from TrueSkill than Elo for a variety of reasons, some being that TrueSkill has better native support for ties (which are more common in our other Simulation competition, ConnectX) and multiplayer games.
+> [Kaggle Staff](https://www.kaggle.com/c/halite/discussion/160083)
+
+The [leaderboard package](leaderboard) generates a homemade leaderboard based on the TrueSkill algorithm:
+
+Agent | TrueSkill | W/L/D | Baseline
+--- | --- | --- | ---
+Negamax ply-4 | 32.22 | 71/7/14 | ✔
+Negamax ply-5 | 30.03 | 45/10/17 | ✔
+Negamax ply-2 | 26.41 | 31/14/19 | ✔
+Negamax ply-1 | 22.34 | 17/45/15 | ✔
+Random | 18.87 | 1/75/17 | ✔
+
+## How to compute the leaderboard?
+To compute the leaderboard:
+```
+cd leaderboard
+python make_leaderboard.py 
+```
+This will compute the leaderboard based on the current [REGISTERED_AGENTS](leaderboard/agents/__init__.py) configuration.
+
+`sigma_threshold` is a parameter that controls the trade-off between precision of the leaderboard values and speed of computation (smaller: more precision. bigger: faster).
```

Comment:
I don't exactly know to be honest. Sigma is defined as the "The degree of uncertainty in the gamer’s skill" in the TrueSkill system: https://www.microsoft.com/en-us/research/project/trueskill-ranking-system/
In theory, this should represent the standard deviation in the agent skill estimate:
![trueskill-skilldia 1](https://user-images.githubusercontent.com/30185046/103481440-f98a6680-4dd2-11eb-818f-1384bf3c928e.jpg)
But, is that really the case...? I don't know. I haven't dived into TrueSkill theory to have an answer.

In practice I think it depends. If with sigma_threshold=1. agents at the top of the leaderboard have a really really close TrueSkill value, so maybe we should consider using a smaller threshold. If this isn't the case, the value should be fine.

Also, another reference in terms of practical value is the test file that I have written: https://github.com/j-bernardi/rl-camb-kaggle-connect-x/blob/1f9293172ee1e3748816a4cc2502f87a8a8bc2b0/leaderboard/trueskill_ranking/test_trueskill_ranking.py
I have written two tests:

- One with determistic results: agent "TheBest" always beats agent "NotBad, who always beats agent "Ex",...

- One with strongly non-deterministic results, the formula is hard to read but basically: "TheBest" wins 5/9 of the time versus "NotBad" and draw 1/9 of the time,...

On both tests I checked that the leaderboard was in the same order every time (TheBest > NotBad > Ex = Aequo > CouldDoBetter > Loser). To enforce this condition, I have put sigma_threshold=2.0 for the deterministic test and sigma_threshold=0.5 for the non-deterministic one (sigma_threshold=1.0 was failing sometimes). With those values of sigma_threshold tests never failed once.
**Conclusion**: if we want to be on the safe side maybe we should consider putting the threshold to 0.5 because this value seems pretty reliable on the test. But honestly 1.0 isn't that bad either and is a lot faster to compute...

Reply:
Thanks for the thorough reply - makes sense that sigma was 1 S.D. - didn't realise that! Seems fine for now then, but good to have this information documented :-)